### PR TITLE
Feature/distribute snipper joiner

### PR DESCRIPTION
**Pull request by hkernbach**
added optimizer rule with distribute node optimizations

### CODE REVIEW

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/9912/

File: arangod/Aql/OptimizerRule.h [hunk 1]
```diff
@@ -269,6 +269,10 @@ struct OptimizerRule {
     // put it after Scatter rule because we would do
     // the work twice otherwise
     removeSatelliteJoinsRule,
+
+    // remove multiple remote <-> distribute snippets if we are able
```
Comment (markuspf) [inline arangod/Aql/OptimizerRule.h:273]:
```suggestion
    // remove multiple gather <-> distribute snippets if we are able to
```

File: arangod/Aql/OptimizerRules.h [hunk 2]
```diff
@@ -168,6 +168,10 @@ void scatterSatelliteGraphRule(Optimizer*, std::unique_ptr<ExecutionPlan>,
 void removeSatelliteJoinsRule(Optimizer*, std::unique_ptr<ExecutionPlan>,
                               OptimizerRule const&);
 
+/// @brief remove distribute/gather and remote nodes if possible
+void removeDistributeNodesRule(Optimizer*, std::unique_ptr<ExecutionPlan>,
+                             OptimizerRule const&);
```
Comment (markuspf) [inline arangod/Aql/OptimizerRules.h:173]:
```suggestion
                               OptimizerRule const&);
```

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/9915/

File: arangod/Aql/OptimizerRulesFeature.cpp [hunk 3]
```diff
@@ -389,6 +389,11 @@ void OptimizerRulesFeature::addRules() {
                OptimizerRule::removeSatelliteJoinsRule,
                OptimizerRule::makeFlags(OptimizerRule::Flags::CanBeDisabled,
                                         OptimizerRule::Flags::ClusterOnly));
+
+  registerRule("remove-distribute-nodes", removeDistributeNodesRule,
```
Comment (joerg84) [inline arangod/Aql/OptimizerRulesFeature.cpp:393]: Do we have a list of rules in docs anywhere?

Comment (markuspf) [reply arangod/Aql/OptimizerRulesFeature.cpp:393]: https://github.com/arangodb/docs/blob/master/3.7/aql/execution-and-performance-optimizer.md (and there needs to be a PR to docs that documents this rule)

**Pull request merged**